src/usdt/mod.rs

```diff
@@ -0,0 +1,187 @@
+use std::{collections::HashSet, ffi::{CStr, CString}, path::{Path, PathBuf}, ptr};
+use std::os::unix::ffi::OsStrExt;
+
+use bcc_sys::bccapi::{bcc_usdt_enable_fully_specified_probe, bcc_usdt_enable_probe, bcc_usdt_genargs, bcc_usdt_get_location, bcc_usdt_location, pid_t};
+use bcc_sys::bccapi::{bcc_usdt_new_frompath, bcc_usdt_new_frompid, bcc_usdt_close};
+use libc::{c_int, c_void};
+
+use crate::{BPF, Uprobe, error::BccError};
+
+struct USDTProbe {
+    binary: String,
+    handler: String,
+    address: u64,
+    pid: Option<pid_t>,
+}
+
+/// Generates the C code for parsing the arguments of all enabled probes in each context.
+pub fn usdt_generate_args(mut contexts: Vec<USDTContext>) -> Result<(String, Vec<USDTContext>), BccError> {
+    // Build an array of pointers to each underlying USDT context.
+    let mut ptrs = Vec::new();
+    for context in &mut contexts {
```

Comment:
Might be able to do `let ptrs = contexts.iter_mut().map(|c| c.as_context_ptr()).collect()` here instead?